cylinder/workers/de/update_de.go

```diff
@@ -0,0 +1,198 @@
+			}
+		}
+	}
+}
+
+// Stop stops the UpdateDE worker.
+func (u *UpdateDE) Stop() error {
+	u.logger.Info("stop")
+	return u.client.Stop()
+}
+
+// subscribe subscribes to the events that trigger the DE update.
+func (u *UpdateDE) subscribe() (err error) {
+	assignedDEQuery := fmt.Sprintf(
+		"%s.%s = '%s'",
+		types.EventTypeRequestSignature,
+		types.AttributeKeyAddress,
+		u.context.Config.Granter,
+	)
+
+	u.eventCh, err = u.client.Subscribe("AssignedDE", assignedDEQuery, 1000)
+	return err
+}
+
+// updateDE updates DE if the remaining DE is too low.
+func (u *UpdateDE) updateDE(numNewDE uint64) {
+	canUpdate, err := u.canUpdateDE()
+	if err != nil {
+		u.logger.Error(":cold_sweat: Cannot update DE: %s", err)
+		return
+	}
+	if !canUpdate {
+		u.logger.Debug(
+			":cold_sweat: Cannot update DE: the granter is not a member of the current or incoming group and gas price isn't set in the config",
+		)
+		return
+	}
+
+	u.logger.Info(":delivery_truck: Updating DE")
+
+	// Generate new DE pairs
+	privDEs, err := GenerateDEs(
+		numNewDE,
+		u.context.Config.RandomSecret,
+		u.context.Store,
+	)
+	if err != nil {
+		u.logger.Error(":cold_sweat: Failed to generate new DE pairs: %s", err)
+		return
+	}
+
+	// Store all DEs in the store
+	var pubDEs []types.DE
+	for _, privDE := range privDEs {
+		pubDEs = append(pubDEs, privDE.PubDE)
+
+		if err := u.context.Store.SetDE(privDE); err != nil {
+			u.logger.Error(":cold_sweat: Failed to set new DE in the store: %s", err)
+			return
+		}
+
+		metrics.IncOffChainDELeftGauge()
+	}
+
+	u.logger.Info(":white_check_mark: Successfully generated %d new DE pairs", numNewDE)
+
+	// Send MsgDE
+	u.context.MsgCh <- types.NewMsgSubmitDEs(pubDEs, u.context.Config.Granter)
+}
+
+// canUpdateDE checks if the system allows to update DEs into the system and chain.
+func (u *UpdateDE) canUpdateDE() (bool, error) {
+	gasPrices, err := sdk.ParseDecCoins(u.context.Config.GasPrices)
+	if err != nil {
+		u.logger.Debug(":cold_sweat: Failed to parse gas prices from config: %s", err)
+	}
+
+	// If the gas price is non-zero, it indicates that the user is willing to pay
+	// a transaction fee for submitting DEs to the chain.
+	if gasPrices != nil && !gasPrices.IsZero() {
+		return true, nil
+	}
+
+	// If the address is a member of the current group, the system can submit DEs to the chain
+	// without paying gas.
+	resp, err := u.client.QueryMember(u.context.Config.Granter)
+	if err != nil {
+		return false, fmt.Errorf("failed to query member information: %w", err)
+	}
+
+	if resp.CurrentGroupMember.Address == u.context.Config.Granter ||
+		resp.IncomingGroupMember.Address == u.context.Config.Granter {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// intervalUpdateDE updates DE on the chain so that the remaining DE is
+// always above the minimum threshold.
+func (u *UpdateDE) intervalUpdateDE() error {
+	deCount, err := u.getDECount()
+	if err != nil {
+		return err
+	}
+
+	if deCount < 2*u.context.Config.MinDE {
```

Comment:
why 2*u.context.Config.MinDE not u.context.Config.MinDE

Reply:
Roger said that the number of DEs on-chain should be at least u.context.Config.MinDE and should not exceed 2 * u.context.Config.MinDE. However, i rewrite to be a new variable so it is easy to read